### PR TITLE
dal-online FS web service integration

### PR DESCRIPTION
**Pull request by talebzeghmi**
- Supports writes w/ hacky method
- Avro schema in Catalog

### CODE REVIEW

File: intake_dal/dal_online.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        """
+
+        self._urlpath = urlpath
+        parse_result = urlparse(urlpath)  # type: ParseResult
+        (self._url, _) = urldefrag(urlpath)
+        self._key_name = parse_result.fragment
+        assert self._key_name != '', "key_name expected in URL fragment"
+        self._key_value = key
+        self._canonical_name = None  # _get_schema() sets this
+        super(DalOnlineSource, self).__init__(metadata=metadata)
```
Comment (smatzana) [inline intake_dal/dal_online.py]: are we supporting Python 2.x as well?  Python 3.x can just have `super()`

File: intake_dal/dal_online.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    def _get_schema(self):
+        if self._canonical_name is None:
```
Comment (smatzana) [inline intake_dal/dal_online.py]: so this method needs to get called in every `write()`? Can it be instantiated once (once metadata is set by parent class I guess?)

File: intake_dal/dal_online.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        )
+
+    def write(self, df: pd.DataFrame):
+        self._get_schema()
+        avro_str = serialize_panda_df_to_str(df, self._avro_schema)
+        self._http_put_avro_data_set(
+            {'data_set_name': self._canonical_name, 'key_name': self._key_name, 'avro_rows': avro_str}
+        )
+
+    def _get_partition(self, _) -> pd.DataFrame:
```
Comment (smatzana) [inline intake_dal/dal_online.py]: function name seems to be completely different from what the body indicates (get partition vs deserialize an avro string)?

File: intake_dal/dal_online.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        self._get_schema()
+        return deserialize_avro_str_to_pandas(self._http_get_avro_data_set(), self._avro_schema)
+
+    def _http_put_avro_data_set(self, json: Dict) -> int:
```
Comment (smatzana) [inline intake_dal/dal_online.py]: if we're using type hints might as well put them in most functions?

File: intake_dal/dal_online.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        response = requests.put(urllib.parse.urljoin(self._url, "avro-data-sets/"), json=json)
+        if response.status_code != 200:
```
Comment (smatzana) [inline intake_dal/dal_online.py]: replace `200` with an HTTP package constant?

File: intake_dal/dal_online.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            raise Exception(f"url={response.url} code={response.status_code}: {response.text}")
+
+        return response.status_code
+
+    def _http_get_avro_data_set(self) -> str:
+        response = requests.get(
+            urllib.parse.urljoin(self._url, f"avro-data-sets/{self._canonical_name}/{self._key_value}"),
```
Comment (smatzana) [inline intake_dal/dal_online.py]: replace `avro-data-sets/` with a const/var as we're referring to it > 2 times hence open to typos/bugs

File: intake_dal/dal_online.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        )
+        if response.status_code != 200:
+            raise Exception(f"url={response.url} code={response.status_code}: {response.text}")
+        return response.json()['avro_rows']
+
+    def _close(self):
+        pass
+
+
+def serialize_panda_df_to_str(df: pd.DataFrame, schema: Dict) -> str:
```
Comment (smatzana) [inline intake_dal/dal_online.py]: move these util functions to different file/package? Or make them `staticmethods` Probably the JVM in me, not sure how pythonic this approach is :)

File: intake_dal/dal_online.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+    ret = {}
+    for (k, v) in field_schemas.items():
+        lookup = to_lookup(v)
+        if lookup in avro_type_to_dtype:
+            ret[k] = avro_type_to_dtype[lookup]
+        elif 'null' in lookup:
+            list_lookup = list(lookup)
+            list_lookup.remove('null')
+            new_lookup = tuple(list_lookup)
```
Comment (smatzana) [inline intake_dal/dal_online.py]: nit: easier

```
new_lookup = tuple(l for l in lookup if l != 'null')
```
```

File: intake_dal/dal_online.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
+from urllib.parse import urlparse, ParseResult, urldefrag  # noqa: F401
```
Comment (smatzana) [inline intake_dal/dal_online.py]: nit: my flake8 places this right after line 7

File: intake_dal/dal_online.py (hunk 3 above)
Comment (talebzeghmi) [reply intake_dal/dal_online.py]: This function name is implementing an Intake base class abstract method.

File: intake_dal/dal_online.py (hunk 7 above)
Comment (talebzeghmi) [reply intake_dal/dal_online.py]: There is a todo or JIRA task to put these Avro methods in a library.

File: intake_dal/dal_online.py (hunk 2 above)
Comment (talebzeghmi) [reply intake_dal/dal_online.py]: it only runs if `self._canonical_name is None`

**Pull request merged**